src/gleam/httpc.gleam

```
@@ -98,6 +136,161 @@ pub fn send_bits(
+          req.method,
+          erl_req,
+          erl_http_options,
+          erl_options,
+        )
+      }
+      _ -> {
+        let erl_content_type =
+          req
+          |> request.get_header("content-type")
+          |> result.unwrap("application/octet-stream")
+          |> charlist.from_string
+        let erl_req = #(erl_url, erl_headers, erl_content_type, req.body)
+        erl_request_async(req.method, erl_req, erl_http_options, erl_options)
+      }
+    }
+    |> result.map_error(normalise_error),
+  )
+
+  Ok(request_id)
+}
+
+// Receive a message that has been sent the the current process using a
+// `NamedSubject`. In this case the subject name is `http`
+@external(erlang, "gleam_httpc_ffi", "receive")
+fn receive(req_id: Reference, timeout: Int) -> Result(#(Atom, Dynamic), Atom)
+
+/// Receive an ansynchronous stream message to the process designated by the
+/// `request_id`.  For a description of those messages, see the documentation for
+/// `httpc.async_stream`. The Dynamic data response may be decoded by the caller
+/// using the convenience functions `httpc.dynamic_to_header_pid`,
+/// `httpc.dynamic_to_header`, and `httpc.dynamic_to_bit_array`
+/// TODO: Refine Error
+pub fn async_receive(
+  req_id: Reference,
+  timeout: Int,
+) -> Result(#(Atom, Dynamic), HttpError) {
+  use response <- result.try(
+    receive(req_id, timeout)
+    |> result.map_error(with: fn(reason) {
+      case atom.to_string(reason) {
+        "timeout" -> ResponseTimeout
+        _ -> InvalidUtf8Response
+      }
+    }),
+  )
+  Ok(response)
+}
+
+///  Call this function when using the `Stream` option `StreamOnce`
+///  It triggers the next message to be sent to the calling process designated by
+/// `pid`. 
+pub fn async_stream_next(pid: Pid) -> Result(Nil, HttpError) {
+  use _next <- result.try(
+    // TODO return reason somehow? {http, {ReqId, {error, Reason}}}
+    stream_next(pid) |> result.replace_error(InvalidUtf8Response),
```

This doesn't seem like the right error value?

---

Yes, you're right.  Actually, with Erlang's `httpc:stream_next/1`, if the Pid is not a valid HTTP client handler (i.e., not managing a live streaming request), then the call will silently fail with no effect. It will still return ok, but no message will be sent, and no chunk will arrive.   I decided to handle this by returning `Result(Nil, Nil)`, regardless of the impossibility of an error.